### PR TITLE
cookbooks: Use new node_search_with_cache helper

### PR DESCRIPTION
**Pull request by vuntz**
~~Requires https://github.com/crowbar/crowbar-core/pull/906 (merged)~~

### CODE REVIEW

File: chef/cookbooks/nova/recipes/compute_ha.rb [hunk 1]
```diff
@@ -35,7 +35,7 @@
 internal_auth_url_v2 = \
   "#{keystone_settings["protocol"]}://" \
   "#{keystone_settings["internal_url_host"]}:#{keystone_settings["service_port"]}/v2.0/"
-neutrons = search(:node, "roles:neutron-server AND roles:neutron-config-#{nova[:nova][:neutron_instance]}")
+neutrons = node_search_with_cache("roles:neutron-server")
```
Comment (vuntz) [inline chef/cookbooks/nova/recipes/compute_ha.rb]: So here it doesn't matter because we have `nova[:nova][:neutron_instance]} == node[:nova][:neutron_instance]}` by definition (node and nova are both part of the same nova proposal, so with the same attribute), so simply using the "magic" behavior will work.

File: chef/cookbooks/swift/recipes/proxy.rb [hunk 2]
```diff
@@ -277,8 +277,7 @@
 end
 
 ## make sure to fetch ring files from the ring compute node
-env_filter = " AND swift_config_environment:#{node[:swift][:config][:environment]}"
-compute_nodes = search(:node, "roles:swift-ring-compute#{env_filter}")
+compute_nodes = node_search_with_cache("roles:swift-ring-compute")
```
Comment (vuntz) [inline chef/cookbooks/swift/recipes/proxy.rb]: Happily ignoring :-)

File: chef/cookbooks/swift/recipes/ring-compute.rb [hunk 3]
```diff
@@ -48,7 +46,7 @@
 
 ####
 # collect the current contents of the ring files.
-nodes = search(:node, "roles:swift-storage#{env_filter}")
+nodes = node_search_with_cache("roles:swift-storage")
```
Comment (vuntz) [inline chef/cookbooks/swift/recipes/ring-compute.rb]: Ignoring, as per https://github.com/crowbar/crowbar-openstack/pull/666/files#r94943772

File: chef/cookbooks/neutron/recipes/cisco_apic_agents.rb [hunk 4]
```diff
@@ -16,8 +16,7 @@
 
 neutron = nil
 if node.attribute?(:cookbook) && node[:cookbook] == "nova"
-  neutrons = search(:node, "roles:neutron-server AND \
-                    roles:neutron-config-#{node[:nova][:neutron_instance]}")
+  neutrons = node_search_with_cache("roles:neutron-server", node[:nova][:neutron_instance])
```
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/cisco_apic_agents.rb]: As discussed on jangouts, yes, they are :-)

**Comment by vuntz** [comment]
Rebased

File: chef/cookbooks/neutron/recipes/common_config.rb [hunk 5]
```diff
@@ -20,8 +20,11 @@
 if node.attribute?(:cookbook) and node[:cookbook] == "nova"
   neutrons = node_search_with_cache("roles:neutron-server", node[:nova][:neutron_instance])
   neutron = neutrons.first || raise("Neutron instance '#{node[:nova][:neutron_instance]}' for nova not found")
+  nova = node
```
Comment (matelakat) [inline chef/cookbooks/neutron/recipes/common_config.rb]: At this point we know that `node[:cookbook] == "nova"`, and based on that information, we say that this is the nova controller. Is that always true?

File: chef/cookbooks/ceilometer/recipes/mongodb.rb [hunk 6]
```diff
@@ -82,9 +82,9 @@
     # replicaset
     package("ruby#{node["languages"]["ruby"]["version"].to_f}-rubygem-mongo").run_action(:install)
 
-    members = search(:node,
-                     "ceilometer_ha_mongodb_replica_set_member:true AND "\
-                     "ceilometer_config_environment:#{node[:ceilometer][:config][:environment]}")
+    members = node_search_with_cache("roles:ceilometer-server").select do |n|
+      n[:ceilometer][:ha][:mongodb][:replica_set][:member] rescue false
```
Comment (matelakat) [inline chef/cookbooks/ceilometer/recipes/mongodb.rb]: Seems like it's the same query as above. Maybe a candidate for extracting?

File: chef/cookbooks/ceilometer/recipes/server.rb [hunk 7]
```diff
@@ -23,10 +23,9 @@
   mongodb_nodes = nil
 
   if ha_enabled
-    mongodb_nodes = search(:node,
-                           "ceilometer_ha_mongodb_replica_set_member:true AND "\
-                           "ceilometer_config_environment:#{node[:ceilometer][:config][:environment]}"
-      )
+    mongodb_nodes = node_search_with_cache("roles:ceilometer-server").select do |n|
+      n[:ceilometer][:ha][:mongodb][:replica_set][:member] rescue false
```
Comment (matelakat) [inline chef/cookbooks/ceilometer/recipes/server.rb]: Same here

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 5 above)
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: No, it's not a nova controller. It's a nova node, but nova nodes have the attributes that we need too.

File: chef/cookbooks/neutron/recipes/cisco_apic_agents.rb (hunk 4 above)
Comment (aspiers) [inline chef/cookbooks/neutron/recipes/cisco_apic_agents.rb]: @vuntz For the benefit of those of us who missed the jangouts discussion, presumably this is because `neutron_config_environment:foo` is Chef magic for searching for nodes which have `node[:neutron][:config][:environment]` set to `foo`.  And if a node has a neutron barclamp proposal `foo` applied to it then it will have both this and also the role `neutron-config-foo`, which is where that attribute comes from.  Right?

Comment (vuntz) [inline chef/cookbooks/neutron/recipes/cisco_apic_agents.rb]: @aspiers yes

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 5 above)
Comment (aspiers) [inline chef/cookbooks/neutron/recipes/common_config.rb]: I found this pretty confusing (not least because apparently this is using a very old Dell hack to provide the `include_recipe` caller via `node[:cookbook]`), and the commit message gives very little context.

So please can you expand the commit message and rename the variables to make this clearer?  It's not immediately obvious what variables called `neutron` and `nova` actually refer to.  I suggest `s/neutrons/neutron_servers/`, `s/neutron/neutron_server/`, and `s/nova/nova_node/` (not so sure about the last one; can you think of anything a bit more informative?)

File: chef/cookbooks/ceilometer/recipes/common.rb [hunk 8]
```diff
@@ -5,10 +5,9 @@
   db_connection = nil
 
   if node[:ceilometer][:ha][:server][:enabled]
-    db_hosts = search(:node,
-                      "ceilometer_ha_mongodb_replica_set_member:true AND roles:ceilometer-server AND "\
-                      "ceilometer_config_environment:#{node[:ceilometer][:config][:environment]}"
-                     )
+    db_hosts = node_search_with_cache("roles:ceilometer-server").select do |n|
+      n[:ceilometer][:ha][:mongodb][:replica_set][:member] rescue false
```
Comment (aspiers) [inline chef/cookbooks/ceilometer/recipes/common.rb]: By now you must be familiar with the reasons for Hound complaining about this. So are you deliberately disagreeing with them, or just old habits dying hard?

Comment (vuntz) [inline chef/cookbooks/ceilometer/recipes/common.rb]: I simply disagree about avoiding `rescue` when we access an attribute with a deep path that may not exist.

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 5 above)
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: > It's not immediately obvious what variables called neutron and nova actually refer to. I suggest s/neutrons/neutron_servers/, s/neutron/neutron_server/, and s/nova/nova_node/ (not so sure about the last one; can you think of anything a bit more informative?)

That would definitely not be part of this PR according to the way you split commits ;-)

File: chef/cookbooks/ceilometer/recipes/mongodb.rb (hunk 6 above)
Comment (vuntz) [inline chef/cookbooks/ceilometer/recipes/mongodb.rb]: @matelakat I added a commit for this.

File: chef/cookbooks/neutron/recipes/common_config.rb (hunk 5 above)
Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: > So please can you expand the commit message and rename the variables to make this clearer?

Done.

Comment (vuntz) [inline chef/cookbooks/neutron/recipes/common_config.rb]: > It's not immediately obvious what variables called neutron and nova actually refer to. I suggest s/neutrons/neutron_servers/, s/neutron/neutron_server/, and s/nova/nova_node/ (not so sure about the last one; can you think of anything a bit more informative?)

@aspiers I started doing that, but I feel this is actually way too prone to errors to be part of this PR. I can create a PR after this one just for this, though.

Comment (aspiers) [inline chef/cookbooks/neutron/recipes/common_config.rb]: > That would definitely not be part of this commit according to the way you split commits ;-)

Correct - if you look closely, you'll see that I wasn't asking the renaming to be part of this commit :-)  Ideally commits refactoring for clarity should happen first, so that when it comes to the non-refactoring commits, they already benefit from the added clarity.

But I'm OK with this being done in a separate PR.  And thanks for the much improved commit message!

**Comment by vuntz** [comment]
@aspiers is this fine with you with the latest changes?

**Comment by vuntz** [comment]
Rebased

**Comment by vuntz** [comment]
HA job found a bug, fixed.

**Comment by vuntz** [comment]
@dirkmueller @rsalevsky @matelakat are your +1 still valid?

**Comment by vuntz** [comment]
Rebased: aodh changes are gone now that the aodh barclamp has been split (and doesn't use mongodb anymore).

**Comment by aplanas** [comment]
Retrying HA job

**Comment by aspiers** [comment]
@dirkmueller I think this one is good to merge now.

(and #toomanyreviewers)
